File: backend/routers/manage.py
"""Post-deploy campaign management endpoints."""
import json
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from pydantic import BaseModel
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from database import get_db
from models.campaign import Project, AdSet, Ad
from models.creative import Creative
from services.campaign_manager import CampaignManager
from services.meta_api import MetaAPIService
from services.claude_creative_analyzer import analyze_creative, analyze_carousel
from routers.auth import get_meta_token, get_meta_auth
from config import get_settings
import os

router = APIRouter(prefix="/api/manage", tags=["manage"])


class BudgetUpdate(BaseModel):
    daily_budget: Optional[float] = None
    lifetime_budget: Optional[float] = None


class AdSetCreateRequest(BaseModel):
    name: str
    targeting_json: str
    optimization_goal: str = "OFFSITE_CONVERSIONS"
    bid_strategy: str = "LOWEST_COST_WITHOUT_CAP"
    budget: Optional[float] = None
    placements_json: Optional[str] = None


class AdCreateRequest(BaseModel):
    name: str
    creative_id: int
    headline: str = ""
    primary_text: str = ""
    description: str = ""
    cta: str = "SHOP_NOW"
    url: Optional[str] = None


@router.put("/{project_id}/budget")
async def update_campaign_budget(
    project_id: int, data: BudgetUpdate, db: AsyncSession = Depends(get_db)
):
    """Update campaign budget (CBO)."""
    auth = await get_meta_auth(db)
    manager = CampaignManager(db, MetaAPIService(auth.access_token, page_id=auth.page_id or ""))
    return await manager.update_budget(project_id, data.daily_budget, data.lifetime_budget)


@router.put("/adset/{adset_id}/budget")
async def update_adset_budget(
    adset_id: int, daily_budget: float, db: AsyncSession = Depends(get_db)
):
    """Update ad set budget (ABO)."""
    auth = await get_meta_auth(db)
    manager = CampaignManager(db, MetaAPIService(auth.access_token, page_id=auth.page_id or ""))
    return await manager.update_adset_budget(adset_id, daily_budget)


@router.post("/{project_id}/adset")
async def add_adset_to_campaign(
    project_id: int, data: AdSetCreateRequest, db: AsyncSession = Depends(get_db)
):
    """Add a new ad set to a deployed campaign."""
    auth = await get_meta_auth(db)
    manager = CampaignManager(db, MetaAPIService(auth.access_token, page_id=auth.page_id or ""))
    adset = await manager.add_adset(
        project_id=project_id,
        name=data.name,
        targeting_json=data.targeting_json,
        optimization_goal=data.optimization_goal,
        bid_strategy=data.bid_strategy,
        budget=data.budget,
        placements_json=data.placements_json,
    )
    return {
        "status": "created",
        "adset_id": adset.id,
        "meta_adset_id": adset.meta_adset_id,
        "generated_name": adset.generated_name,
    }


@router.post("/adset/{adset_id}/ad")
async def add_ad_to_adset(
    adset_id: int, data: AdCreateRequest, db: AsyncSession = Depends(get_db)
):
    """Add a new ad to an existing ad set."""
    auth = await get_meta_auth(db)
    manager = CampaignManager(db, MetaAPIService(auth.access_token, page_id=auth.page_id or ""))
    ad = await manager.add_ad(
        adset_id=adset_id,
        name=data.name,
        creative_id=data.creative_id,
        headline=data.headline,
        primary_text=data.primary_text,
        description=data.description,
        cta=data.cta,
        url=data.url,
    )
    return {
        "status": "created",
        "ad_id": ad.id,
        "meta_ad_id": ad.meta_ad_id,
        "generated_name": ad.generated_name,
    }


@router.post("/{project_id}/analyze-creative")
async def analyze_new_creative(
    project_id: int,
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
):
    """
    Upload and analyze a new creative with AI.
    Returns recommendation for which ad set it should be placed in.
    User must confirm before any action is taken.
    """
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")

    # Save uploaded file
    settings = get_settings()
    upload_dir = os.path.join(settings.storage_path, str(project_id), "new_creatives")
    os.makedirs(upload_dir, exist_ok=True)
    file_path = os.path.join(upload_dir, file.filename)

    with open(file_path, "wb") as f:
        content = await file.read()
        f.write(content)

    # Determine media type
    ext = os.path.splitext(file.filename)[1].lower()
    media_type = "video" if ext in {".mp4", ".mov", ".avi", ".mkv"} else "image"

    # Get campaign strategy from parsed structure
    strategy = ""
    if project.parsed_structure_json:
        parsed = json.loads(project.parsed_structure_json)
        strategy = f"Campaign: {parsed.get('campaign', {}).get('name', project.name)}\n"
        strategy += f"Objective: {project.campaign_objective}\n"
        strategy += "Ad Sets (each represents a different angle/audience):\n"
        for adset_data in parsed.get("ad_sets", []):
            strategy += f"- {adset_data.get('name', 'Unknown')}"
            if adset_data.get("description"):
                strategy += f": {adset_data['description']}"
            strategy += "\n"

    # Get existing ad sets
    adsets_result = await db.execute(
        select(AdSet).where(AdSet.project_id == project_id)
    )
    ad_sets = [
        {"id": a.id, "name": a.name, "description": a.generated_name}
        for a in adsets_result.scalars().all()
    ]

    # Analyze with AI
    try:
        analysis = await analyze_creative(file_path, media_type, strategy, ad_sets)
    except Exception as e:
        raise HTTPException(500, f"Creative analysis failed: {e}")

    # Save creative to DB (but don't assign to any ad set yet - user must confirm)
    from services.google_drive import GoogleDriveService
    creative = Creative(
        project_id=project_id,
        original_name=file.filename,
        base_name=os.path.splitext(file.filename)[0],
        media_type=media_type,
        local_path=file_path,
        file_size_bytes=os.path.getsize(file_path),
    )
    db.add(creative)
    await db.commit()
    await db.refresh(creative)

    return {
        "creative_id": creative.id,
        "filename": file.filename,
        "media_type": media_type,
        "analysis": analysis,
        "message": "Review the AI recommendation and confirm the placement.",
    }


class StatusUpdate(BaseModel):
    status: str  # "ACTIVE" or "PAUSED"


@router.put("/ad/{ad_id}/status")
async def update_ad_status(
    ad_id: int, data: StatusUpdate, db: AsyncSession = Depends(get_db)
):
    """Toggle ad status (ACTIVE/PAUSED)."""
    result = await db.execute(select(Ad).where(Ad.id == ad_id))
    ad = result.scalar_one_or_none()
    if not ad or not ad.meta_ad_id:
        raise HTTPException(404, "Ad not found or not deployed")

    auth = await get_meta_auth(db)
    meta = MetaAPIService(auth.access_token, page_id=auth.page_id or "")
    await meta.update_ad_status(ad.meta_ad_id, data.status)

    ad.status = data.status.lower()
    await db.commit()

    return {"status": "updated", "ad_id": ad.id, "new_status": data.status}


class CopyUpdate(BaseModel):
    headline: Optional[str] = None
    primary_text: Optional[str] = None
    description: Optional[str] = None


@router.put("/ad/{ad_id}/copy")
async def update_ad_copy(
    ad_id: int,
    data: CopyUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update ad copy (local DB only - requires re-deploy to sync to Meta)."""
    result = await db.execute(select(Ad).where(Ad.id == ad_id))
    ad = result.scalar_one_or_none()
    if not ad:
        raise HTTPException(404, "Ad not found")

    if data.headline is not None:
        ad.headline = data.headline
    if data.primary_text is not None:
        ad.primary_text = data.primary_text
    if data.description is not None:
        ad.description = data.description

    await db.commit()
    return {"status": "updated", "ad_id": ad.id}

File: api/lib/routes/manage.ts
import { Router, Request, Response } from "express";
import { eq } from "drizzle-orm";
import multer from "multer";
import fs from "fs";
import path from "path";
import { db } from "../db/client.js";
import { projects, adSets, ads, creatives } from "../db/schema.js";
import { getMetaAuth } from "./auth.js";

const upload = multer({ dest: "/tmp/uploads" });
const router = Router();

// ── PUT /:id/budget ───────────────────────────────────────────────────────

router.put("/:id/budget", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id as string, 10);
    const { daily_budget, lifetime_budget } = req.body;

    const auth = await getMetaAuth(db);

    const { CampaignManager } = await import("../services/campaign-manager.js");
    const { MetaApiService } = await import("../services/meta-api.js");

    const meta = new MetaApiService(auth.accessToken, auth.pageId ?? "");
    const manager = new CampaignManager(db, meta);

    const result = await manager.updateBudget(projectId, daily_budget, lifetime_budget);
    return res.json(result);
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── PUT /adset/:id/budget ─────────────────────────────────────────────────

router.put("/adset/:id/budget", async (req: Request, res: Response) => {
  try {
    const adsetId = parseInt(req.params.id as string, 10);
    const { daily_budget } = req.body;

    if (daily_budget == null) {
      return res.status(400).json({ detail: "daily_budget is required" });
    }

    const auth = await getMetaAuth(db);

    const { CampaignManager } = await import("../services/campaign-manager.js");
    const { MetaApiService } = await import("../services/meta-api.js");

    const meta = new MetaApiService(auth.accessToken, auth.pageId ?? "");
    const manager = new CampaignManager(db, meta);

    const result = await manager.updateAdSetBudget(adsetId, daily_budget);
    return res.json(result);
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── POST /:id/adset ──────────────────────────────────────────────────────

router.post("/:id/adset", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id as string, 10);
    const {
      name,
      targeting_json,
      optimization_goal = "OFFSITE_CONVERSIONS",
      bid_strategy = "LOWEST_COST_WITHOUT_CAP",
      budget,
      placements_json,
    } = req.body;

    if (!name) {
      return res.status(400).json({ detail: "name is required" });
    }

    const auth = await getMetaAuth(db);

    const { CampaignManager } = await import("../services/campaign-manager.js");
    const { MetaApiService } = await import("../services/meta-api.js");

    const meta = new MetaApiService(auth.accessToken, auth.pageId ?? "");
    const manager = new CampaignManager(db, meta);

    const adset = await manager.addAdSet(
      projectId,
      name,
      targeting_json,
      optimization_goal,
      bid_strategy,
      budget ?? null,
      placements_json ?? null,
    );

    return res.json({
      status: "created",
      adset_id: adset.id,
      meta_adset_id: adset.metaAdsetId,
      generated_name: adset.generatedName,
    });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── POST /adset/:id/ad ───────────────────────────────────────────────────

router.post("/adset/:id/ad", async (req: Request, res: Response) => {
  try {
    const adsetId = parseInt(req.params.id as string, 10);
    const {
      name,
      creative_id,
      headline = "",
      primary_text = "",
      description = "",
      cta = "SHOP_NOW",
      url,
    } = req.body;

    if (!name) {
      return res.status(400).json({ detail: "name is required" });
    }
    if (!creative_id) {
      return res.status(400).json({ detail: "creative_id is required" });
    }

    const auth = await getMetaAuth(db);

    const { CampaignManager } = await import("../services/campaign-manager.js");
    const { MetaApiService } = await import("../services/meta-api.js");

    const meta = new MetaApiService(auth.accessToken, auth.pageId ?? "");
    const manager = new CampaignManager(db, meta);

    const ad = await manager.addAd(
      adsetId,
      name,
      creative_id,
      headline,
      primary_text,
      description,
      cta,
      url ?? null,
    );

    return res.json({
      status: "created",
      ad_id: ad.id,
      meta_ad_id: ad.metaAdId,
      generated_name: ad.generatedName,
    });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── PUT /ad/:id/status ────────────────────────────────────────────────────

router.put("/ad/:id/status", async (req: Request, res: Response) => {
  try {
    const adId = parseInt(req.params.id as string, 10);
    const { status } = req.body as { status: string };

    if (!status || !["ACTIVE", "PAUSED"].includes(status.toUpperCase())) {
      return res.status(400).json({ detail: "status must be ACTIVE or PAUSED" });
    }

    const adRows = await db
      .select()
      .from(ads)
      .where(eq(ads.id, adId))
      .limit(1);
    const ad = adRows[0];
    if (!ad || !ad.metaAdId) {
      return res.status(404).json({ detail: "Ad not found or not deployed" });
    }

    const auth = await getMetaAuth(db);

    const { MetaApiService } = await import("../services/meta-api.js");
    const meta = new MetaApiService(auth.accessToken, auth.pageId ?? "");
    await meta.updateAdStatus(ad.metaAdId, status.toUpperCase());

    await db
      .update(ads)
      .set({ status: status.toLowerCase() })
      .where(eq(ads.id, adId));

    return res.json({ status: "updated", ad_id: ad.id, new_status: status.toUpperCase() });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── PUT /ad/:id/copy ──────────────────────────────────────────────────────

router.put("/ad/:id/copy", async (req: Request, res: Response) => {
  try {
    const adId = parseInt(req.params.id as string, 10);
    const { headline, primary_text, description } = req.body;

    const adRows = await db
      .select()
      .from(ads)
      .where(eq(ads.id, adId))
      .limit(1);
    const ad = adRows[0];
    if (!ad) {
      return res.status(404).json({ detail: "Ad not found" });
    }

    const updates: Record<string, unknown> = {};
    if (headline !== undefined) updates.headline = headline;
    if (primary_text !== undefined) updates.primaryText = primary_text;
    if (description !== undefined) updates.description = description;

    if (Object.keys(updates).length > 0) {
      await db.update(ads).set(updates).where(eq(ads.id, adId));
    }

    return res.json({ status: "updated", ad_id: ad.id });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── POST /:id/analyze-creative ────────────────────────────────────────────

router.post(
  "/:id/analyze-creative",
  upload.single("file"),
  async (req: Request, res: Response) => {
    try {
      const projectId = parseInt(req.params.id as string, 10);

      const projectRows = await db
        .select()
        .from(projects)
        .where(eq(projects.id, projectId))
        .limit(1);
      const project = projectRows[0];
      if (!project) {
        return res.status(404).json({ detail: "Project not found" });
      }

      const file = req.file;
      if (!file) {
        return res.status(400).json({ detail: "No file uploaded" });
      }

      // Save uploaded file
      const uploadDir = path.join("/tmp/uploads", String(projectId), "new_creatives");
      fs.mkdirSync(uploadDir, { recursive: true });
      const filePath = path.join(uploadDir, file.originalname);
      fs.renameSync(file.path, filePath);

      // Determine media type
      const ext = path.extname(file.originalname).toLowerCase();
      const VIDEO_EXTS = new Set([".mp4", ".mov", ".avi", ".mkv"]);
      const mediaType = VIDEO_EXTS.has(ext) ? "video" : "image";

      // Get campaign strategy from parsed structure
      let strategy = "";
      if (project.parsedStructureJson) {
        const parsed = JSON.parse(project.parsedStructureJson);
        strategy += `Campaign: ${parsed.campaign?.name ?? project.name}\n`;
        strategy += `Objective: ${project.campaignObjective}\n`;
        strategy += "Ad Sets (each represents a different angle/audience):\n";
        for (const adsetData of parsed.ad_sets ?? []) {
          strategy += `- ${adsetData.name ?? "Unknown"}`;
          if (adsetData.description) {
            strategy += `: ${adsetData.description}`;
          }
          strategy += "\n";
        }
      }

      // Get existing ad sets
      const adsetRows = await db
        .select()
        .from(adSets)
        .where(eq(adSets.projectId, projectId));

      const adSetList = adsetRows.map((a) => ({
        id: a.id,
        name: a.name,
        description: a.generatedName,
      }));

      // Analyze with AI
      let analysis: any;
      try {
        const { analyzeCreative } = await import("../services/claude-creative-analyzer.js");
        analysis = await analyzeCreative(filePath, mediaType, strategy, adSetList);
      } catch (e: any) {
        return res.status(500).json({ detail: `Creative analysis failed: ${e.message}` });
      }

      // Save creative to DB (unassigned - user must confirm placement)
      const creativeResult = await db
        .insert(creatives)
        .values({
          projectId,
          originalName: file.originalname,
          baseName: path.parse(file.originalname).name,
          mediaType: mediaType as "image" | "video",
          localPath: filePath,
          fileSizeBytes: fs.statSync(filePath).size,
          uploadStatus: "downloaded",
        })
        .returning();
      const creative = creativeResult[0];

      return res.json({
        creative_id: creative.id,
        filename: file.originalname,
        media_type: mediaType,
        analysis,
        message: "Review the AI recommendation and confirm the placement.",
      });
    } catch (e: any) {
      return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
    }
  },
);

export default router;
